azure/src/main/java/org/apache/iceberg/azure/adlsv2/AzureSasCredentialRefresher.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.azure.adlsv2;
+
+import com.azure.core.credential.AzureSasCredential;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class AzureSasCredentialRefresher {
+  private final Supplier<Pair<String, Long>> sasTokenWithExpirationSupplier;
+  private final ScheduledExecutorService refreshExecutor;
+  private final AzureSasCredential azureSasCredential;
+
+  private static final long MAX_REFRESH_WINDOW_MILLIS = 300_000; // 5 minutes;
+  private static final long MIN_REFRESH_WAIT_MILLIS = 10;
+
+  public AzureSasCredentialRefresher(
+      Supplier<Pair<String, Long>> sasTokenWithExpirationSupplier,
```

Comment:
does this actually need to be a supplier given that it's being immediately used in L40?

Reply:
Yes, the supplier is being used during initialization and also during each scheduled refresh to fetch the new credentials. Supplier holds the logic to fetch new credentials from the API endpoint, since we are going to use it multiple times I modelled it as supplier instead of single method call. Please suggest if there is a cleaner way to achieve the same.